### PR TITLE
Capture kernel failure information after we have analysed the errors

### LINKED ISSUE #10990: Kernel failures are not properly identifiable
When kernel startup fails due to missing ipykernel we detect them as just `kernel died` errors instead of an ipykernel not found error
In the past we'd detect them correctly

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #10990

### CODE REVIEW

File: src/kernels/errors/kernelErrorHandler.ts [hunk 1]
```diff
@@ -222,23 +224,35 @@ export abstract class DataScienceErrorHandler implements IDataScienceErrorHandle
 
         // Jupyter kernels, non zmq actually do the dependency install themselves
         if (isCancellationError(err)) {
+            traceVerbose(`Cancelled by user`);
+            this.sendKernelTelemetry(err, errorContext, resource, 'cancelled');
             return KernelInterpreterDependencyResponse.cancel;
         } else if (err instanceof JupyterKernelDependencyError) {
             traceWarning(`Jupyter Kernel Dependency Error, reason=${err.reason}`, err);
-            if (err.reason === KernelInterpreterDependencyResponse.uiHidden) {
+            this.sendKernelTelemetry(err, errorContext, resource, err.category);
```
Comment (DonJayamanne) [inline src/kernels/errors/kernelErrorHandler.ts:232]: Now send telemetry after we've analysed the error

File: src/kernels/errors/kernelErrorHandler.ts [hunk 2]
```diff
@@ -328,15 +344,12 @@ export abstract class DataScienceErrorHandler implements IDataScienceErrorHandle
                 })
                 .then(noop, noop);
             return KernelInterpreterDependencyResponse.failed;
-        } else if (isCancellationError(err)) {
```
Comment (DonJayamanne) [inline src/kernels/errors/kernelErrorHandler.ts:331]: Duplicaate, we already do this at the start of this if condition, see line 226

File: src/kernels/errors/kernelErrorHandler.ts [hunk 3]
```diff
@@ -222,23 +224,35 @@ export abstract class DataScienceErrorHandler implements IDataScienceErrorHandle
...
+            this.sendKernelTelemetry(err, errorContext, resource, err.category);
+            if (err.reason === KernelInterpreterDependencyResponse.uiHidden && this.kernelDependency) {
                 // At this point we're handling the error, and if the error was initially swallowed due to
                 // auto start (ui hidden), now we need to display the error to the user.
-                const response = this.dependencyManager
-                    ? await this.dependencyManager.installMissingDependencies(err)
-                    : JupyterInterpreterDependencyResponse.cancel;
-                return response === JupyterInterpreterDependencyResponse.ok
-                    ? KernelInterpreterDependencyResponse.ok
-                    : KernelInterpreterDependencyResponse.cancel;
```
Comment (DonJayamanne) [inline src/kernels/errors/kernelErrorHandler.ts:236]: This was wrong, if ipykernel is missing, then we shouldn't be trying to install jupyter dependencies
Possible users were running into this

File: src/kernels/errors/kernelErrorHandler.ts [hunk 4]
```diff
@@ -306,6 +321,7 @@ export abstract class DataScienceErrorHandler implements IDataScienceErrorHandle
             }
             return KernelInterpreterDependencyResponse.cancel;
         } else if (err instanceof JupyterSelfCertsError) {
+            this.sendKernelTelemetry(err, errorContext, resource, err.category);
```
Comment (DonJayamanne) [inline src/kernels/errors/kernelErrorHandler.ts:324]: Kernel Telemetry (for failrues) is sent in this class after we've analysed the failure.

File: src/kernels/raw/session/rawJupyterSession.node.ts [hunk 5]
```diff
@@ -106,17 +106,6 @@ export class RawJupyterSession extends BaseJupyterSession implements IRawKernelC
                 sendKernelTelemetryEvent(this.resource, Telemetry.RawKernelSessionStartUserCancel);
                 traceVerbose('Starting of raw session cancelled by user');
                 throw error;
-            } else if (error instanceof TimedOutError) {
```
Comment (DonJayamanne) [inline src/kernels/raw/session/rawJupyterSession.node.ts:109]: `TimeoutError` class was never used anywhere, meaning we'd never throw such an error anywhere.

File: src/kernels/serviceRegistry.node.ts [hunk 6]
```diff
@@ -109,8 +108,6 @@ export function registerTypes(serviceManager: IServiceManager, isDevMode: boolea
     // This will ensure all subsequent telemetry will get the context of whether it is a custom/native/old notebook editor.
     // This is temporary, and once we ship native editor this needs to be removed.
     setSharedProperty('ds_notebookeditor', 'native');
-    const isLocalConnection = serviceManager.get<IServerConnectionType>(IServerConnectionType).isLocalLaunch;
-    setSharedProperty('localOrRemoteConnection', isLocalConnection ? 'local' : 'remote');
```
Comment (DonJayamanne) [inline src/kernels/serviceRegistry.node.ts:113]: This was not correct anymore, as we can have both local and remote connectsion together

File: src/platform/common/utils/async.ts [hunk 7]
```diff
@@ -3,20 +3,9 @@
...
 
-import { BaseError } from '../../errors/types';
-
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type PromiseFunction = (...any: any[]) => Promise<any>;
 
-/**
- * Error type thrown when a timeout occurs
- */
-export class TimedOutError extends BaseError {
```
Comment (DonJayamanne) [inline src/platform/common/utils/async.ts:14]: Never used anywhere

File: src/platform/api/pythonApi.ts [hunk 8]
```diff
@@ -324,14 +325,18 @@ export class InterpreterService implements IInterpreterService {
...
     public getInterpreters(resource?: Uri): Promise<PythonEnvironment[]> {
         this.hookupOnDidChangeInterpreterEvent();
         // Cache result as it only changes when the interpreter list changes or we add more workspace folders
+        const firstTime = !!this.interpretersFetchedOnceBefore;
+        this.interpretersFetchedOnceBefore = true;
         if (!this.interpreterListCachePromise) {
             this.interpreterListCachePromise = this.getInterpretersImpl(resource);
         }
+        sendTelemetryWhenDone(Telemetry.InterpreterListingPerf, this.interpreterListCachePromise, undefined, {
+            firstTime
```
Comment (DonJayamanne) [inline src/platform/api/pythonApi.ts:338]: This telemetry was lost in the refactoring, we need a way to check how long it takes to fetch the list of interpreters for the first time in the session.

File: src/kernels/errors/kernelErrorHandler.ts [hunk 9]
```diff
@@ -222,23 +224,35 @@ export abstract class DataScienceErrorHandler implements IDataScienceErrorHandle
 
         // Jupyter kernels, non zmq actually do the dependency install themselves
         if (isCancellationError(err)) {
+            traceVerbose(`Cancelled by user`);
```
Comment (amunger) [inline src/kernels/errors/kernelErrorHandler.ts:227]: This wouldn't be very clear in a log file, can you add more context - even just a tag like [kernelErrorHandler]

Comment (DonJayamanne) [reply src/kernels/errors/kernelErrorHandler.ts]: Remvoed that, i think its unnecessary.

**Review by amunger — APPROVED** [review]

**Pull request merged**